test_mapbased_param/Cl_estimation.py

```diff
@@ -0,0 +1,375 @@
+        ### compute power spectra of the cleaned sky maps
+        ncomp = int(len(clean_map)/2)
+        Cl_clean = [ell_eff] 
+        Cl_noise = [ell_eff] 
+        components = []
+        print('n_comp = ', ncomp)
+        ind=0
+        for comp_i in range(ncomp):
+            # for comp_j in range(ncomp)[comp_i:]:
+            comp_j = comp_i
+            print('comp_i = ', comp_i)
+            print('comp_j = ', comp_j)
+
+            components.append(str((comp_i,comp_j))) 
+
+            ## signal spectra
+            if comp_i > 1: purify_b_=False
+            else: purify_b_=True
+            fyp_i=get_field(mask*clean_map[2*comp_i], mask*clean_map[2*comp_i+1], mask_apo, purify_b=purify_b_)
+            # fyp_i=get_field(clean_map[2*comp_i], clean_map[2*comp_i+1], purify_b=purify_b_)
+            fyp_j=get_field(mask*clean_map[2*comp_j], mask*clean_map[2*comp_j+1], mask_apo, purify_b=purify_b_)
+            # fyp_j=get_field(clean_map[2*comp_j], clean_map[2*comp_j+1], purify_b=purify_b_)
+            Cl_clean.append(compute_master(fyp_i, fyp_j, w)[3])
+
+            ## noise spectra
+            # fyp_i_noise=get_field(mask*post_compsep_noise[2*comp_i], mask*post_compsep_noise[2*comp_i+1], purify_b=True)
+            # fyp_i_noise=get_field(mask_nh*post_compsep_noise[2*comp_i], mask_nh*post_compsep_noise[2*comp_i+1], purify_b=True)
+            fyp_i_noise=get_field(mask*post_compsep_noise[2*comp_i], mask*post_compsep_noise[2*comp_i+1], mask_apo, purify_b=True)
+            # fyp_i_noise=get_field(post_compsep_noise[2*comp_i], post_compsep_noise[2*comp_i+1], purify_b=True)
+            # fyp_j_noise=get_field(mask*post_compsep_noise[2*comp_j], mask*post_compsep_noise[2*comp_j+1], purify_b=True)
+            # fyp_j_noise=get_field(mask_nh*post_compsep_noise[2*comp_j], mask_nh*post_compsep_noise[2*comp_j+1], purify_b=True)
+            fyp_j_noise=get_field(mask*post_compsep_noise[2*comp_j], mask*post_compsep_noise[2*comp_j+1], mask_apo, purify_b=True)
+            # fyp_j_noise=get_field(post_compsep_noise[2*comp_j], post_compsep_noise[2*comp_j+1], purify_b=True)
+
+            Cl_noise.append(compute_master(fyp_i_noise, fyp_j_noise, w)[3])
+
+            ind += 1
+
+        print('ind = ', ind)
+        print('shape(Cl_clean) = ', len(Cl_clean))
+        print('shape(array(Cl_clean)) = ', (np.array(Cl_clean)).shape)
+        print('all components = ', components)
+        print('saving to disk ... ')
+        hp.fitsfunc.write_cl(self.get_output('Cl_clean'), np.array(Cl_clean), overwrite=True)
+        hp.fitsfunc.write_cl(self.get_output('Cl_noise'), np.array(Cl_noise), overwrite=True)
+        hp.fitsfunc.write_cl(self.get_output('Cl_cov_clean'), np.array(Cl_cov_clean), overwrite=True)
+        hp.fitsfunc.write_cl(self.get_output('Cl_cov_freq'), np.array(Cl_cov_freq), overwrite=True)
+        hp.fitsfunc.write_cl(self.get_output('Cl_noise_bias'), np.array(Cl_noise_bias), overwrite=True)
```

Comment:
So, at some point I think you should start using `sacc`. The main reason for this is that later on you're gonna want to convolve the theory with the same bandpower window functions that are in the data, and `sacc` allows you to save that concurrently with the power spectra and the covariance easily. Especially with purification at low ell, the bandpower convolution can be quite important to avoid biases.